### ISSUE TITLE
Change Table::ReadRows to work with StatusOr.

### ISSUE THREAD

**Issue by coryan** [googleapis/google-cloud-cpp#2001]
The `google::cloud::bigtable::Table::ReadRows()` function return a `bigtable::RowReader`.  This class has `begin()` and `end()` member functions that return an iterator of `bigtable::Row`.

The iterator (which is of type `google::cloud::RowReader::iterator`, an alias for `google::cloud::bigtable::internal::RowReaderIterator`), raises an exception (which can be disabled, but that is not the case we are worried about here).  Instead of raising exceptions, it should become an iterator over `StatusOr<Row>`.

At least, I think that is how we will fix this class to support `StatusOr`, I would like @devjgm to confirm.

**Comment by devjgm** [comment]
LGTM

**Comment by dopiera** [comment]
Greg and I already started the discussion on the iterator in the Spanner DD. Let me move it here.

# TL;DR;
 I'm not a big fan of iterator over `StatusOr<>` because we're introducing another iterator state (in case of an error we get an iterator which is not `end()` but indicates and end of iteration). That disallows using such iterators in standard algorithms. Instead I propose a facade, which binds the status:
```C++
Status status;
for (auto& row : table.ReadRows(...).iter(status)) {
   // handle rows
}
if (!status.ok()) {
  // oh no stream was broken...
}
```

# Alternatives
I agree it's a bit awkward, but TBH, I thought it's the least bad idea I considered. Let me document my thought process:

## Idea 1
You'd have to check status on result_set after you've finished iterating and it would tell you about an error condition while receiving the stream.

### Snippet
```C++
auto row_reader = table.ReadRows(...);
for (auto& row : row_reader) {
   // handle rows
}
if (!row_reader.stream_status().ok()) {
  // oh no stream was broken
}
```
### Pros:
* range-based loops work
* not very tedious to use
### Cons:
* user might forget to check the status at the end

## Idea 2:
Rather than implementing a typical C++ iterator, create one which has `HasNext()` / `Next()` member functions (like Java iterators). `HasNext()` could return a Status.

### Snippet
```C++
auto it = table.ReadRows(...).iterator()
StatusOr<bool> has_next;
while (has_next = it.HasNext()).ok() && *has_next) {
   auto& row = it.Next();
   // handle rows
}
if (!has_next) {
  // oh no stream was broken, handle has_next.status()
}
```
### Pros:
* pretty straight forward
* hard to forget about the status
### Cons:
* very non-idiomatic
* range-based loops no longer work

### Idea 3: (proposal from the description):
Make the iterator return `StatusOr<>`

### Snippet
```C++
for (auto& row : table.ReadRows(...)) {
   if (!row) {
     // oh no stream was broken, handle row.status()
     break;
   }
   // handle rows
}
```

### Pros:
- impossible to miss the stream status
- range-based loops almost work (you need to manually break them)
### Cons:
- weird semantics - we'd introduce a third state of an iterator (beyond valid an invalid): it's fine to dereference it, but doesn't hold value; it's not end(), but indicates an end
- you wouldn't be able to do a `std::copy` or any other standard algorithm

## Idea 4 (AKA my proposal):
Create a facade around ResultSet, which binds the status, so that if the stream is broken, the iterators advances to end() and the bound status is set accordingly.

### Snippet
```C++
Status status;
for (auto& row : table.ReadRows(...).iter(status)) {
   // handle rows
}
if (!status.ok()) {
  // oh no stream was broken...
}
```

### Pros:
- almost impossible to miss the status
- as idiomatic as it gets with StatusOr
- all the assumptions around iterators still work, you can use `std::copy` and company.
### Cons:
- it's a bit unusual to check errors that way

What do you think?

**Comment by coryan** [comment]
I am not a fan of iterators of `StatusOr<>` either, but FWIW, we considered all these options (or slight variations of them) when designing the GCS iterators. We picked Idea 3 *after* considering them.

@devjgm: we should write that discussion as a design doc or ADR, we should not be rehashing it on every iteration API.

I think we all agree that iterators with `StatusOr<>` are never going to be as clean as iterators with exceptions (but we cannot do that), and non-iterator APIs are worse.

If it helps, do not think of `StatusOr<>` as an error for the iterator, the iteration *happens* to stop on the first error, but that is part of the range.  Just like if you had a collection of `std::optional<T>` and it just happened that the last optional had no value :unamused: 

Alternatively: it is not hard to write a wrapper that converts the iterator of `StatusOr<T>` to an iterator that raises on the first non-ok  `StatusOr<>`.

In any case, my comment is not binding here, we should wait for  @devjgm to chime in.

**Comment by devjgm** [comment]
Thanks for your detailed write up @dopiera. While `StatusOr` certainly does make things a bit more clumsy, Idea 3 is the one that I prefer. I agree with @ocryan's explanation about an alternative way to think about an iterator of StatusOr that makes it, IMO, not so bad. Idea 3 also fits well with all the other cases where we *return* Status or StatusOr, i.e., we never (I think never?) return them via an out-param.

**Comment by dopiera** [comment]
Thanks, I'll move on with option 3. I wasn't aware we already did that in GCS.
